File: Same Binary Tree - Leetcode 100/Same Binary Tree - Leetcode 100.cpp
#include <algorithm>

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        return isSame(p, q);
    }

private:
    bool isSame(TreeNode* p, TreeNode* q) {
        if (!p && !q) return true;
        if (!p || !q) return false;
        if (p->val != q->val) return false;

        return isSame(p->left, q->left) && isSame(p->right, q->right);
    }
};

File: Same Binary Tree - Leetcode 100/Same Binary Tree - Leetcode 100.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # 1: Both Null
        if not p and not q: 
            return True
        
        # 2: One is Null
        if (p and not q) or (q and not p):
            return False
        
        # 3. Values Mismatch
        if p.val != q.val:
            return False
        
        return self.isSameTree(p.left, q.left) and \
               self.isSameTree(p.right, q.right)

# Time Complexity: O(n)
# Space Complexity: O(h) { here "h" is the height of the tree }
